pkg/resources/server_test.go

```diff
@@ -0,0 +1,327 @@
+			rs := resourceServer{rsConnector: rsc}
+			err := rs.Start()
+			Expect(err).To(HaveOccurred())
+		})
+		It("start server with plugin failed to dial server", func() {
+			grpcServer := grpc.NewServer([]grpc.ServerOption{}...)
+			rsc := &mocks.ResourceServerConnector{}
+			rsc.On("GetServer", mock.Anything).Return(grpcServer)
+			rsc.On("CreateServer").Return()
+			rsc.On("Listen", mock.Anything, mock.Anything).Return(nil, nil)
+			rsc.On("Serve", mock.Anything).Return()
+			rsc.On("Dial", mock.Anything, mock.Anything).Return(nil, errors.New("failed"))
+			rs := resourceServer{rsConnector: rsc}
+			err := rs.Start()
+			Expect(err).To(HaveOccurred())
+		})
+		It("start server with plugin watcher disabled failed to register", func() {
+			grpcServer := grpc.NewServer([]grpc.ServerOption{}...)
+			rsc := &mocks.ResourceServerConnector{}
+			rsc.On("GetServer", mock.Anything).Return(grpcServer)
+			rsc.On("CreateServer").Return()
+			rsc.On("Listen", mock.Anything, mock.Anything).Return(nil, nil)
+			rsc.On("Serve", mock.Anything).Return()
+			rsc.On("Dial", mock.Anything, mock.Anything).Return(nil, nil)
+			rsc.On("Close", mock.Anything).Return()
+			rsc.On("Stop").Return()
+			rsc.On("Register", mock.Anything, mock.Anything).Return(errors.New("failed"))
+			rs := resourceServer{watchMode: false, rsConnector: rsc}
+			err := rs.Start()
+			Expect(err).To(HaveOccurred())
+		})
+	})
+	Context("Stop", func() {
+		It("stop server with correct parameters", func() {
+			grpcServer := grpc.NewServer([]grpc.ServerOption{}...)
+			rsc := &mocks.ResourceServerConnector{}
+			rsc.On("DeleteServer").Return()
+			rsc.On("GetServer").Return(grpcServer)
+			rsc.On("Stop").Return()
+
+			stopWatcher := make(chan bool)
+			rs := resourceServer{
+				rsConnector: rsc,
+				stopWatcher: stopWatcher,
+				stop:        make(chan interface{}),
+			}
+			// Dummy listener to stopWatcher to not block the test and fail
+			go func() {
+				<-stopWatcher
+			}()
+
+			err := rs.Stop()
+			Expect(err).ToNot(HaveOccurred())
+		})
+		It("stop non existing server", func() {
+			rs := resourceServer{}
+			err := rs.Stop()
+			Expect(err).ToNot(HaveOccurred())
+		})
+	})
+	Context("Restart", func() {
+		It("Restart server with correct parameters", func() {
+			grpcServer := grpc.NewServer([]grpc.ServerOption{}...)
+			rsc := &mocks.ResourceServerConnector{}
+			rsc.On("GetServer", mock.Anything).Return(grpcServer)
+			rsc.On("CreateServer").Return()
+			rsc.On("DeleteServer").Return()
+			rsc.On("Listen", mock.Anything, mock.Anything).Return(nil, nil)
+			rsc.On("Serve", mock.Anything).Return()
+			rsc.On("Dial", mock.Anything, mock.Anything).Return(nil, nil)
+			rsc.On("Stop").Return()
+			rsc.On("Close", mock.Anything).Return()
+
+			rs := resourceServer{
+				watchMode:   true,
+				rsConnector: rsc,
+				stop:        make(chan interface{}),
+			}
+
+			err := rs.Restart()
+			Expect(err).ToNot(HaveOccurred())
```

Comment:
any way to ensure Stop() and Start() where called ?

Reply:
also a negative test case is missing here

Reply:
To ensure Stop and Start calls added the AssertExpectations, also added negative case